src/services/invitation/service.ts

```diff
@@ -136,4 +140,98 @@ export class InvitationService {
     const memberships = invitations.map(({ permission, item }) => ({ item, member, permission }));
     await itemMembershipRepository.createMany(memberships);
   }
+
+  async handleCSVInvitations(
+    actor: Actor,
+    repositories: Repositories,
+    query: IdParam,
+    file: MultipartFile | undefined,
+  ) {
+    const { id: parentId, template_id } = query;
+    const parentItem = await repositories.itemRepository.get(parentId);
+    await validatePermission(repositories, PermissionLevel.Admin, actor, parentItem);
+
+    if (!file || file.mimetype != CSV_MIMETYPE) {
+      throw new Error(`
+          An incorrect type of file has been uploaded, 
+          Please upload a file with .csv extension
+        `);
+    }
+
+    const stream = file.file;
+    const { rows, header } = await getCSV(stream);
+    const hasGrpCol = header.includes(GRP_COL_NAME);
+    let groupNameToItem = new Map<string, Item>();
+
+    if (hasGrpCol) {
+      groupNameToItem = await rows.reduce(
+        async (result: Promise<Map<string, Item>>, rowInvite: CSVInvite) => {
+          const r = await result;
+          if (!r.has(rowInvite.group_name as string)) {
+            const itemCreated = await this.itemService.post(actor, repositories, {
+              item: {
+                name: rowInvite.group_name,
+                type: ItemType.FOLDER,
+              },
+              parentId: parentId,
+            });
+
+            if (template_id) {
+              //get the all items of from template_id and copy them to new folder
+              const itemsInTemplate = await this.itemService.getDescendants(
+                actor,
+                repositories,
+                template_id,
+              );
+              const regex = regexGen(
+                (await this.itemService.get(actor, repositories, template_id)).path,
+              );
+
+              const itemsAtFirstLevel = itemsInTemplate
+                .filter((tmpItem) => {
+                  return regex.test(tmpItem.path);
+                })
+                .map((tmpItem) => {
+                  return tmpItem.id;
+                });
+
+              await this.itemService.copyMany(actor, repositories, itemsAtFirstLevel, {
+                parentId: itemCreated.id,
+              });
+            }
+            r.set(rowInvite.group_name as string, itemCreated);
+          }
+          return r;
+        },
+        Promise.resolve(new Map<string, Item>()),
+      );
```

Comment:
Please refactor this reduce to create an execution plan synchronously (no await) and then perform the async actions. You could create a map with some options that tels you what you will have to do. Also add comments to explain why you do certain things as their purpose is not easy to understand.